### ISSUE TITLE
static_casting to a large unsigned long long from a double causes an FPE

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#18060]
|  |  |
| --- | --- |
| Bugzilla Link | [17686](https://llvm.org/bz17686) |
| Version | trunk |
| OS | Linux |
| Attachments | [Sample code that causes an unexpected FPE](https://user-images.githubusercontent.com/60944935/143748346-91eb40ec-0ead-4895-b1d3-34f067f79fa1.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @isanbard,@jayfoad,@RKSimon,@zygoloid,@vinc17fr,@vlad902 |
| Fixed by commit(s) | 348251 |

## Extended Description 

I am getting an FPE when converting a double to an unsigned long long, if the double is larger than the max (signed) long long.  It is as if the static_cast code is using the signed integer internally, for the conversion from a double to the int type, then preforms a second conversion from the signed to unsigned.
```c
double x = 9.223372036854776e+18;
unsigned long long ullX = static_cast<unsigned long long>(x);
```
I've attached a full example that shows the problem when running clang-3.3 on Linux.  The code works various versions of gcc that I tried.

**Comment by vinc17fr** [comment]
Under Debian, I still get the issue up to Debian clang version 11.1.0-4+b2, but it has disappeared in Debian clang version 12.0.1-17+b1. If this has been fixed, perhaps the issue should be closed.

**Comment by RKSimon** [comment]
I'm not sure what could have fixed this - https://reviews.llvm.org/D89697 improved FPTOUI conversion using the SSE instructions 'min signed value' overflow result, but iirc we'd still be raising fp exceptions if they were enabled?

**Comment by Rombur** [comment]
> iirc we'd still be raising fp exceptions if they were enabled?

That's right. I am still hitting this bug with Debian clang version 19.1.7 if I add `feenableexcept(FE_INVALID);`

This passes:
```
double x = 9.223372036854776e+18;
unsigned long long ullX = static_cast<unsigned long long>(x);
```

but this fails:
```
feenableexcept(FE_INVALID);
double x = 9.223372036854776e+18;
unsigned long long ullX = static_cast<unsigned long long>(x);
```

It works fine with gcc.

**Comment by vinc17fr** [comment]
What actually changed between Clang 11.1.0 and Clang 12.0.1 is the behavior on the following C code:
```c
#include <stdio.h>
#include <stdint.h>
#include <fenv.h>
#pragma STDC FENV_ACCESS ON

int main (void)
{
  uint64_t i = (uint64_t) 1 << 63;
  volatile double d = i;
  feclearexcept (FE_INVALID);
  int t1 = (uint64_t) d != i;
  int t2 = fetestexcept (FE_INVALID);
  if (t1 || t2)
    printf ("Error in cast of double to unsigned: %s value%s\n",
            t1 ? "incorrect" : "correct", t2 ? ", FE_INVALID" : "");
  return 0;
}
```

With Clang 11 on x64_64, I get:
```
Error in cast of double to unsigned: correct value, FE_INVALID
```

I get no errors with Clang 12 to 19.

**Comment by RKSimon** [comment]
Closing old ticket - from Clang 12.00 onward we correctly use strict\constrained uitofp/fptoui conversions when we detect `#pragma STDC FENV_ACCESS ON`
